src/time.rs

```diff
@@ -0,0 +1,137 @@
+use std::fmt::Write;
+use std::time::Duration;
+
+pub fn format_duration(duration: &Duration) -> String {
+    let total_seconds = duration.as_secs();
+
+    let hours = total_seconds / 3600;
+    let minutes = (total_seconds % 3600) / 60;
+    let seconds = total_seconds % 60;
+
+    let hours_label = match hours {
+        1 => "hour",
+        _ => "hours"
+    };
+
+    let minutes_label = match minutes {
+        1 => "minute",
+        _ => "minutes"
+    };
+
+    let seconds_label = match seconds {
+        1 => "second",
+        _ => "seconds"
+    };
+
+    let mut text = String::new();
+
+    if hours > 0 {
+        push_value_and_label(&mut text, hours, hours_label);
+    }
+
+    if minutes > 0 {
+        if !text.is_empty() {
+            text.push(' ');
+        }
+
+        push_value_and_label(&mut text, minutes, minutes_label);
+    }
+
+    if !text.is_empty() {
+        text.push(' ');
+    }
+
+    push_value_and_label(&mut text, seconds, seconds_label);
```

Comment:
You are still joining strings yourself...

Reply:
?